### ISSUE TITLE
implement ehist()

### ISSUE THREAD

**Issue by brendangregg** [iovisor/bpftrace#1120]
I mentioned it in the book (page 186) and have spoken in the past to @richardelling about it, but someone should implement exponential histograms: ehist(val). For cases where you'd like more detail, you can switch hist() to ehist() by adding the "e".

It can be based on exponential histograms by Jim Roskind:

https://chromium.googlesource.com/chromium/src/base/+/master/metrics/histogram.h#12
https://code.fb.com/developer-tools/performance-scale-2016-recap/

those are the ones you see in about:histograms:

```

0     ------------------------------------------------------------------------O (306321 = 56.4%)
1     -------------------------------O                                          (132875 = 24.5%) {56.4%}
2     ---------------O                                                          (65279 = 12.0%) {80.8%}
3     ----O                                                                     (18564 = 3.4%) {92.8%}
4     --O                                                                       (6481 = 1.2%) {96.3%}
5     -O                                                                        (3650 = 0.7%) {97.4%}
6     -O                                                                        (2390 = 0.4%) {98.1%}
7     O                                                                         (1811 = 0.3%) {98.6%}
8     O                                                                         (3163 = 0.6%) {98.9%}
10    O                                                                         (1334 = 0.2%) {99.5%}
12    O                                                                         (615 = 0.1%) {99.7%}
14    O                                                                         (423 = 0.1%) {99.8%}
17    O                                                                         (161 = 0.0%) {99.9%}
20    O                                                                         (92 = 0.0%) {99.9%}
24    O                                                                         (36 = 0.0%) {100.0%}
29    O                                                                         (15 = 0.0%) {100.0%}
34    O                                                                         (10 = 0.0%) {100.0%}
40    O                                                                         (5 = 0.0%) {100.0%}
48    O                                                                         (16 = 0.0%) {100.0%}
57    O                                                                         (15 = 0.0%) {100.0%}
68    O                                                                         (26 = 0.0%) {100.0%}
81    O                                                                         (23 = 0.0%) {100.0%}
96    O                                                                         (22 = 0.0%) {100.0%}
114   O                                                                         (7 = 0.0%) {100.0%}
135   O                                                                         (2 = 0.0%) {100.0%}
160   O                                                                         (10 = 0.0%) {100.0%}
190   O                                                                         (10 = 0.0%) {100.0%}
226   O                                                                         (1 = 0.0%) {100.0%}
268   O                                                                         (12 = 0.0%) {100.0%}
318   O                                                                         (1 = 0.0%) {100.0%}
378   O                                                                         (8 = 0.0%) {100.0%}
449   O                                                                         (2 = 0.0%) {100.0%}
533   O                                                                         (2 = 0.0%) {100.0%}
633   O                                                                         (2 = 0.0%) {100.0%}
752   O                                                                         (0 = 0.0%) {100.0%}
894   O                                                                         (1 = 0.0%) {100.0%}
1062  O                                                                         (0 = 0.0%) {100.0%}
```

Although I'd make it visually identical to hist() just with different buckets.

Here's my code for lhist() 6bdd1198e04392aa468b12357a051816f2cc50e4 -- but ehist() it's closer to the hist() implementation, as it only has one argument, so lhist() may be overly complex as an example. I can't find the exact commit for just hist().

**Comment by ajor** [comment]
Reading the Chromium page you linked to it looks like they need the min and max values and number of buckets to be specified:

```
// That bucket allocation would actually result from construction of a histogram
// for values between 1 and 64, with 8 buckets, such as:
// Histogram count("some name", 1, 64, 8);
```
We need to define the buckets before we collect the data so bpftrace can't work them out for itself - users would have to specify them like we do for lhist, i.e. `ehist(int n, int min, int max, int nbuckets)`

Another option could be to add an optional argument to `hist`, telling it what to use as the base for its exponential histograms (it's always 2 at the moment). e.g. `@x = hist(arg0, 1.5);` would give us buckets like these (depending on rounding implementation):
```
[0]
[1]
[2,3)
[3,5)
[5,7)
[7,11)
[11,17)
[17,25)
[25,38)
...
```

**Comment by brendangregg** [comment]
I like the look of:

```
@x = hist(arg0);        // power of 2
@x = hist(arg0, 1.5);   // power of 1.5
```

If someone specifies 1.001 it may be limited by max elements (is it the `max_entries = 4096;`?), but if that's a problem we could put a semantic check on it and reject silly power of's. And implement ehist() with the specific bucket count if needed.

**Comment by luigirizzo** [comment]
I have submitted a pull request https://github.com/iovisor/bpftrace/pull/2831 which implements something similar.

Each power of 2 is divided into $2^K$ buckets, with K specified as second argument to `hist(n[, k])` and defaults to 0.
This way we get a good approximation of finer granularity logarithms without the runtime complexitiy (only a couple of shifts and add/mask are needed). This is similar to code I used earlier in  https://github.com/luigirizzo/lr-cstats.
